nomenclature/core.py

```diff
@@ -15,37 +15,41 @@
 class DataStructureDefinition:
     """Definition of datastructure codelists for dimensions used in the IAMC format"""
 
-    def __init__(self, path):
+    def __init__(self, path, dimensions=["region", "variable"]):
         """
 
         Parameters
         ----------
         path : str or path-like
             The folder with the project definitions.
+        dimensions : list of str, optional
+            List of :meth:`CodeList` names, initialized from a sub-folder of `path`.
         """
         if not isinstance(path, Path):
             path = Path(path)
 
         if not path.is_dir():
             raise NotADirectoryError(f"Definitions directory not found: {path}")
 
-        self.variable = CodeList.from_directory("variable", path / "variable")
-        self.region = CodeList.from_directory("region", path / "region")
+        self.dimensions = dimensions
+        for dim in dimensions:
+            self.__setattr__(dim, CodeList.from_directory(dim, path / dim))
 
-        self.dimensions = ["region", "variable"]
         empty = [d for d in self.dimensions if not self.__getattribute__(d)]
         if empty:
             _empty = ", ".join(empty)
```

Comment:
Unrelated to the PR itself but I would say that `_empty` is not needed, the concatenation can be done inline directly in the `ValueError(f"Empty codelist: {', '.join(empty)}")`

Reply:
Done, I had a longer error message earlier which was too long for one line...